### PR TITLE
[CVP] Infer range return attribute

### LINKED ISSUE #98946: function return `range` metadata not inferred from `assume`s
The propagation of range information from `assume` is very fragile. 
Things that prevent it from propagating:

- The `assume` is in its own function
- The condition passed to `assume` is extracted into a helper function (eg `is_surrogate`)
- The condition passed to `assume` combines two comparisons with `or`
- The condition passed to `assume` is the negation of another comparison

https://godbolt.org/z/hvdxE1xKn
```c
#include <stdbool.h>
#include <stdint.h>

typedef uint32_t u32;

void assume(bool b) { __builtin_assume(b); }

bool is_surrogate(u32 c) { return 0xD800 <= c && c <= 0xDFFF; }

// range not inferred
u32 f1(u32 c) {
    assume(!is_surrogate(c));
    return c;
}

// range not inferred
u32 f2(u32 c) {
    __builtin_assume(!is_surrogate(c));
    return c;
}

// range not inferred
u32 f3(u32 c) {
    __builtin_assume(!(0xD800 <= c && c <= 0xDFFF));
    return c;
}

// range not inferred
u32 f4(u32 c) {
    __builtin_assume(c < 0xD800 || c > 0xDFFF);
    return c;
}

// range(i32 0, 55296) inferred
u32 f5(u32 c) {
    __builtin_assume(c < 0xD800);
    return c;
}

// range(i32 55297, 0) inferred
u32 f6(u32 c) {
    __builtin_assume(c > 0xD800);
    return c;
}

// range not inferred
u32 f7(u32 c) {
    __builtin_assume(!(0xD800 <= c));
    return c;
}

// range not inferred
u32 f8(u32 c) {
    __builtin_assume(!(c <= 0xDFFF));
    return c;
}
```

### PR DESCRIPTION
**Pull request by nikic**
We already infer this in IPSCCP (which runs very early, so cannot benefit from inlining and simplifications) and SCCP (which runs without PredicateInfo, so does not use assumes). Do it in CVP as well, so it can handle cases that IPSCCP/SCCP can't. 

Fixes https://github.com/llvm/llvm-project/issues/98946 (everything apart from f2, where the assume is dropped by the frontend).

(No compile-time impact: https://llvm-compile-time-tracker.com/compare.php?from=f1ff3a279f3320d8e0d4abbc1a8357bb51de25a2&to=46ab5f809414b3f06ab9956e0fc6910284417b6d&stat=instructions:u)

### CODE REVIEW

File: llvm/lib/Transforms/Scalar/CorrelatedValuePropagation.cpp [hunk 1]
```diff
@@ -1289,6 +1299,14 @@ static bool runImpl(Function &F, LazyValueInfo *LVI, DominatorTree *DT,
     FnChanged |= BBChanged;
   }
 
+  // Infer range attribute on return value.
+  if (RetRange && !RetRange->isFullSet()) {
+    Attribute RangeAttr = F.getRetAttribute(Attribute::Range);
+    if (RangeAttr.isValid())
+      RetRange = RetRange->intersectWith(RangeAttr.getRange());
+    if (!RetRange->isEmptySet() && !RetRange->isSingleElement())
```
Comment (nikic) [inline llvm/lib/Transforms/Scalar/CorrelatedValuePropagation.cpp:1307]: I've excluded the single element case to make this less spammy. I didn't want to have range attributes on every single function returning a constant (in which case IPSCCP would already propagate the cosntant).

**Comment by dtcxzyw** [comment]
Missing fold: https://alive2.llvm.org/ce/z/2rmc3h
See https://github.com/dtcxzyw/llvm-opt-benchmark/pull/1342#discussion_r1767033125

File: llvm/lib/Transforms/Scalar/CorrelatedValuePropagation.cpp (hunk 1 above)
Comment (goldsteinn) [reply llvm/lib/Transforms/Scalar/CorrelatedValuePropagation.cpp:1307]: Maybe leave that as a comment?

File: llvm/lib/Transforms/Scalar/CorrelatedValuePropagation.cpp [hunk 2]
```diff
@@ -1289,6 +1299,14 @@ static bool runImpl(Function &F, LazyValueInfo *LVI, DominatorTree *DT,
...
+      F.addRangeRetAttr(*RetRange);
```
Comment (goldsteinn) [inline llvm/lib/Transforms/Scalar/CorrelatedValuePropagation.cpp:1308]: Missing `FnChanged = true`?

Comment (nikic) [reply llvm/lib/Transforms/Scalar/CorrelatedValuePropagation.cpp:1310]: Good point, fixed.

**Comment by nikic** [comment]
> Missing fold: https://alive2.llvm.org/ce/z/2rmc3h See [dtcxzyw/llvm-opt-benchmark#1342 (comment)](https://github.com/dtcxzyw/llvm-opt-benchmark/pull/1342#discussion_r1767033125)

In principle, I think this could be supported by making decomposeBitTestICmp detect that `ugt 255` means `& 256 == 1` together with information from computeConstantRange(). Not sure if it's worth it...

**Review by dtcxzyw — APPROVED** [review]
LGTM. Thank you!

File: llvm/lib/Transforms/Scalar/CorrelatedValuePropagation.cpp (hunk 2 above)
Comment (goldsteinn) [reply llvm/lib/Transforms/Scalar/CorrelatedValuePropagation.cpp:1310]: Could we cause an inf loop if existing range attr is same as new one?

**Pull request merged**